Add mutedeck extension

chadrwalters opened the pull request:
## Description

# MuteDeck Extension for Raycast

## Features

- 🎤 **Microphone Control**: Quick mute/unmute across any meeting platform
- 📹 **Video Control**: Toggle camera on/off with a single command
- 🚪 **Meeting Exit**: Leave meetings instantly with confirmation protection
- 📊 **Status Display**: Real-time meeting status with automatic updates
- 🛡 **Smart Protection**: Confirmation dialogs when presenting or recording
- 🔔 **Toast Notifications**: Instant feedback for all actions

## Requirements

- macOS 12.0 or later
- Raycast 1.50.0 or later
- [MuteDeck](https://mutedeck.com) installed and running

## Screenshots

1. **Command List**: Shows all available MuteDeck commands
2. **Status Display**: Real-time meeting status and controls
3. **Confirmation Dialog**: Smart protection while presenting

## Testing

1. Install [MuteDeck](https://mutedeck.com)
2. Install the extension
3. Configure API endpoint (default: http://localhost:3491)
4. Join a meeting in any supported platform
5. Try each command:
   - Toggle microphone
   - Toggle video
   - Show status
   - Leave meeting
6. Test safety features:
   - Start presenting/recording
   - Try toggling controls to see confirmation dialogs

## Privacy & Security

- Communicates only with local MuteDeck instance
- No data sent to external servers
- All preferences stored locally
- No tracking or analytics

## Additional Notes

- Extension follows Raycast's design guidelines
- Implements all recommended safety features
- Provides comprehensive error handling
- Includes detailed documentation

## Screencast
![confirmation](https://github.com/user-attachments/assets/ff9eb9fb-a38d-475e-a680-113011e5e269)
![list-view](https://github.com/user-attachments/assets/68c143cf-6c30-42b3-a2e6-04b5f67aef16)
![status-display](https://github.com/user-attachments/assets/1c579e4d-721c-414b-b230-ae348400ce93)

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

chadrwalters commented:
Thanks @greptileai 

@greptileai I've addressed all the feedback:

1. Configuration:
   - Removed Windsurf-specific ignore patterns
   - Removed unused example file exceptions
   - Removed Swift-related ignores

2. API Implementation:
   - Added proper endpoint validation
   - Improved URL construction using URL class
   - Added timeout handling
   - Enhanced status code checking (2xx range)
   - Optimized preferences caching

3. Type Safety:
   - Added custom error classes (MuteDeckError, MuteDeckTimeoutError, MuteDeckConfigError)
   - Added strict types for API paths and preferences
   - Improved error handling types

4. Documentation:
   - Updated security documentation explaining HTTP usage
   - Removed unnecessary store assets
   - Updated screenshot documentation

5. Code Cleanup:
   - Removed unused mutedeck.tsx
   - Improved error messages
   - Enhanced code organization

Please review the changes.

chadrwalters commented:
@greptileai I've addressed all the feedback:

1. Configuration Cleanup:
   - Removed Windsurf-specific ignore patterns
   - Removed Swift-related ignores
   - Cleaned up example file patterns

2. API Implementation:
   - Using URL constructor for safe URL handling
   - Added timeout handling with AbortController
   - Added proper status code range checking (2xx)
   - Added apiEndpoint validation
   - Cached preferences to avoid duplicate calls
   - Extracted common API call logic into makeApiCall function

3. Code Organization:
   - Created notifications.ts for toast handling
   - Added constants.ts for messages and troubleshooting steps
   - Improved error handling consistency
   - Fixed TypeScript type safety

4. Documentation:
   - Consistent "MuteDeck" naming throughout
   - Updated security documentation
   - Fixed screenshot documentation

Please review the changes when you have a chance. Thanks!

chadrwalters commented:
@greptileai I've addressed the latest feedback:

1. Fixed dependency version issues:
   - Updated dev dependencies to match lock file versions
   - Regenerated package-lock.json for consistent builds

2. Fixed TypeScript errors:
   - Moved `isSuccessStatus` outside of MuteDeckClient class to be globally accessible
   - Fixed access to status code checking in helper functions

3. Improved code organization:
   - Using shared utility functions
   - Fixed code formatting
   - Removed PLAN.md

Please review when you have a chance. Thanks!

chadrwalters commented:
@greptileai I've fixed the timeout handling in MuteDeckClient:

- Made timeout parameter optional in getInstance
- Added logic to recreate instance if timeout changes
- This ensures consistent timeout values across the application

Please review when you have a chance. Thanks!

chadrwalters commented:
@greptileai I've added explicit undefined checking:

- Check for undefined apiTimeout value
- Throw MuteDeckConfigError if preference is missing
- This satisfies TypeScript's type checking
- Still includes NaN and positive value validation

Please review when you have a chance. Thanks!

chadrwalters commented:
@greptileai I've added runtime type validation for the API response:

- Added type guard functions for each status type (ActiveStatus, FeatureStatus, etc.)
- Added comprehensive validation of API response structure
- Throw MuteDeckError with clear message if response format is invalid
- This ensures runtime type safety matches our TypeScript types

Please review when you have a chance. Thanks!

chadrwalters commented:
@greptileai I've added JSON parsing error handling:

- Added try-catch around response.json()
- Throw MuteDeckError with clear message for invalid JSON
- This ensures we handle malformed API responses gracefully

Please review when you have a chance. Thanks!

chadrwalters commented:
@greptileai I've fixed multiple issues:

1. Removed unsafe AbortSignal type cast:
   - Now using signal directly from AbortController
   - TypeScript correctly infers the type

2. Added response validation for toggle operations:
   - New validateToggleResponse method
   - Checks JSON response if present
   - Validates status code for success

3. Added timeout validation:
   - Added MAX_TIMEOUT constant (30 seconds)
   - Clamp timeout values to reasonable range
   - Keep default 5000ms for invalid values

Please review when you have a chance. Thanks!

chadrwalters commented:
@greptileai I've added proper type checking for toggle responses:

- Added StatusResponse interface for API responses
- Added isStatusResponse type guard function
- Removed unsafe type assertion
- This ensures type-safe handling of response status

Please review when you have a chance. Thanks!

pernielsentikaer commented on extensions/mutedeck/CHANGELOG.md:
```diff
@@ -0,0 +1,29 @@
+
+## [Initial Version] - {PR_MERGE_DATE}
+
+## [Next Version]
+
+### Improvements
+- Enhanced type safety throughout the API
+- Improved error handling with custom error classes
+- Removed unnecessary files and cleaned up configuration
+- Updated documentation and screenshots
+- Fixed API endpoint validation
+- Added proper timeout handling
+- Optimized preferences handling
+- Cleaned up store assets
+
+### Technical Updates
+- Added MuteDeckError classes for better error handling
+- Implemented strict TypeScript types
+- Added proper URL validation
+- Improved status code handling
+- Enhanced configuration validation
+- Removed unused placeholder code
+
+### Documentation
+- Updated security documentation
+- Added screenshot specifications
+- Improved API documentation
+- Enhanced error messaging
```
We should not have this in the Changelog, could be moved to README

pernielsentikaer reviewed (commented):
Cam we remove `store/screencast.gif` - 11.6MB is alot of add to the history

pernielsentikaer commented on extensions/mutedeck/package.json:
```diff
@@ -0,0 +1,127 @@
+  "commands": [
+    {
+      "name": "show-status",
+      "title": "Show Status",
```
```suggestion
      "title": "Show Status",
      "subtitle": "Mutedeck",
```

chadrwalters replied:
DOne

pernielsentikaer commented on extensions/mutedeck/package.json:
```diff
@@ -0,0 +1,127 @@
+    },
+    {
+      "name": "toggle-microphone",
+      "title": "Toggle Microphone",
```
```suggestion
      "title": "Toggle Microphone",
      "subtitle": "Mutedeck",
```

chadrwalters replied:
DOne

pernielsentikaer commented on extensions/mutedeck/package.json:
```diff
@@ -0,0 +1,127 @@
+    },
+    {
+      "name": "toggle-video",
+      "title": "Toggle Video",
```
```suggestion
      "title": "Toggle Video",
      "subtitle": "Mutedeck",
```

chadrwalters replied:
DOne

pernielsentikaer commented on extensions/mutedeck/package.json:
```diff
@@ -0,0 +1,127 @@
+    },
+    {
+      "name": "leave-meeting",
+      "title": "Leave Meeting",
```
```suggestion
      "title": "Leave Meeting",
      "subtitle": "Mutedeck",
```

chadrwalters replied:
DOne

pernielsentikaer reviewed (changes requested):
A few more 😊

chadrwalters commented:
Thanks @pernielsentikaer ! I've addressed all your feedback:

1. Moved the detailed changelog content from CHANGELOG.md to a new "Changelog" section in the README.md
2. Simplified the CHANGELOG.md to follow the standard Raycast format
3. Added the "subtitle" field to all commands in package.json
4. Removed the large screencast.gif file (11.6MB) as requested

I've verified there are no references to the screencast file in the codebase, so removing it shouldn't affect any functionality.


Great release this week.

pernielsentikaer commented:
@greptileai can you check again

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥

chadrwalters commented:
Great. Thank you all
